File: MCD-Agencia-main/backend/apps/orders/services/shipping.py
"""Shipping rates from Acapulco, Guerrero to the rest of Mexico."""

from __future__ import annotations

from dataclasses import dataclass
from decimal import Decimal
import re
import unicodedata


def _normalize(value: str) -> str:
    text = unicodedata.normalize('NFD', value or '')
    text = ''.join(ch for ch in text if unicodedata.category(ch) != 'Mn')
    text = text.lower()
    text = re.sub(r'[^a-z0-9\s]', ' ', text)
    return re.sub(r'\s+', ' ', text).strip()


# Official state keys after accent stripping.
STATE_SHIPPING_RATES = {
    'aguascalientes': Decimal('289.00'),
    'baja california': Decimal('499.00'),
    'baja california sur': Decimal('529.00'),
    'campeche': Decimal('369.00'),
    'chiapas': Decimal('329.00'),
    'chihuahua': Decimal('449.00'),
    'ciudad de mexico': Decimal('199.00'),
    'coahuila': Decimal('389.00'),
    'colima': Decimal('229.00'),
    'durango': Decimal('349.00'),
    'estado de mexico': Decimal('199.00'),
    'guanajuato': Decimal('269.00'),
    'guerrero': Decimal('99.00'),
    'hidalgo': Decimal('219.00'),
    'jalisco': Decimal('269.00'),
    'michoacan': Decimal('199.00'),
    'morelos': Decimal('169.00'),
    'nayarit': Decimal('289.00'),
    'nuevo leon': Decimal('379.00'),
    'oaxaca': Decimal('149.00'),
    'puebla': Decimal('189.00'),
    'queretaro': Decimal('249.00'),
    'quintana roo': Decimal('429.00'),
    'san luis potosi': Decimal('299.00'),
    'sinaloa': Decimal('399.00'),
    'sonora': Decimal('459.00'),
    'tabasco': Decimal('319.00'),
    'tamaulipas': Decimal('359.00'),
    'tlaxcala': Decimal('209.00'),
    'veracruz': Decimal('219.00'),
    'yucatan': Decimal('419.00'),
    'zacatecas': Decimal('319.00'),
}

STATE_LABELS = {
    'aguascalientes': 'Aguascalientes',
    'baja california': 'Baja California',
    'baja california sur': 'Baja California Sur',
    'campeche': 'Campeche',
    'chiapas': 'Chiapas',
    'chihuahua': 'Chihuahua',
    'ciudad de mexico': 'Ciudad de México',
    'coahuila': 'Coahuila',
    'colima': 'Colima',
    'durango': 'Durango',
    'estado de mexico': 'Estado de México',
    'guanajuato': 'Guanajuato',
    'guerrero': 'Guerrero',
    'hidalgo': 'Hidalgo',
    'jalisco': 'Jalisco',
    'michoacan': 'Michoacán',
    'morelos': 'Morelos',
    'nayarit': 'Nayarit',
    'nuevo leon': 'Nuevo León',
    'oaxaca': 'Oaxaca',
    'puebla': 'Puebla',
    'queretaro': 'Querétaro',
    'quintana roo': 'Quintana Roo',
    'san luis potosi': 'San Luis Potosí',
    'sinaloa': 'Sinaloa',
    'sonora': 'Sonora',
    'tabasco': 'Tabasco',
    'tamaulipas': 'Tamaulipas',
    'tlaxcala': 'Tlaxcala',
    'veracruz': 'Veracruz',
    'yucatan': 'Yucatán',
    'zacatecas': 'Zacatecas',
}

STATE_ALIASES = {
    'ags': 'aguascalientes',
    'bc': 'baja california',
    'bcs': 'baja california sur',
    'camp': 'campeche',
    'chis': 'chiapas',
    'chih': 'chihuahua',
    'cdmx': 'ciudad de mexico',
    'df': 'ciudad de mexico',
    'd f': 'ciudad de mexico',
    'distrito federal': 'ciudad de mexico',
    'ciudad mexico': 'ciudad de mexico',
    'mexico city': 'ciudad de mexico',
    'coah': 'coahuila',
    'coahuila de zaragoza': 'coahuila',
    'col': 'colima',
    'dgo': 'durango',
    'edomex': 'estado de mexico',
    'edo de mexico': 'estado de mexico',
    'edo mexico': 'estado de mexico',
    'estado mexico': 'estado de mexico',
    'mex': 'estado de mexico',
    'mexico': 'estado de mexico',
    'gto': 'guanajuato',
    'gro': 'guerrero',
    'hgo': 'hidalgo',
    'jal': 'jalisco',
    'mich': 'michoacan',
    'mor': 'morelos',
    'nay': 'nayarit',
    'nl': 'nuevo leon',
    'n l': 'nuevo leon',
    'oax': 'oaxaca',
    'pue': 'puebla',
    'qro': 'queretaro',
    'q roo': 'quintana roo',
    'qroo': 'quintana roo',
    'qr': 'quintana roo',
    'slp': 'san luis potosi',
    'sin': 'sinaloa',
    'son': 'sonora',
    'tab': 'tabasco',
    'tamps': 'tamaulipas',
    'tam': 'tamaulipas',
    'tlax': 'tlaxcala',
    'ver': 'veracruz',
    'veracruz de ignacio de la llave': 'veracruz',
    'yuc': 'yucatan',
    'zac': 'zacatecas',
}

# Sepomex ranges used by Acapulco de Juárez.
ACAPULCO_POSTAL_PREFIXES = {'393', '394', '395', '396', '397', '398', '399'}
DEFAULT_NATIONAL_RATE = Decimal('299.00')


@dataclass(frozen=True)
class ShippingQuote:
    fee: Decimal
    is_free: bool
    is_acapulco: bool
    state_key: str
    state_label: str
    zone_key: str
    label: str

    def to_dict(self) -> dict:
        return {
            'fee': str(self.fee),
            'is_free': self.is_free,
            'is_acapulco': self.is_acapulco,
            'state_key': self.state_key,
            'state_label': self.state_label,
            'zone_key': self.zone_key,
            'label': self.label,
        }


def resolve_state_key(state: str) -> str:
    normalized = _normalize(state)
    if not normalized:
        return ''
    if normalized in STATE_SHIPPING_RATES:
        return normalized
    if normalized in STATE_ALIASES:
        return STATE_ALIASES[normalized]
    for key, label in STATE_LABELS.items():
        if _normalize(label) == normalized:
            return key
    return ''


def is_acapulco_destination(city: str = '', state: str = '', postal_code: str = '') -> bool:
    city_name = _normalize(city)
    if city_name:
        return 'acapulco' in city_name

    digits = re.sub(r'\D', '', postal_code or '')
    if len(digits) >= 5 and digits[:3] in ACAPULCO_POSTAL_PREFIXES:
        state_key = resolve_state_key(state)
        if not state_key or state_key == 'guerrero':
            return True
    return False


def calculate_shipping_quote(
    city: str = '',
    state: str = '',
    postal_code: str = '',
) -> ShippingQuote:
    """Return the shipping fee from Acapulco to a Mexican destination."""
    if is_acapulco_destination(city=city, state=state, postal_code=postal_code):
        return ShippingQuote(
            fee=Decimal('0.00'),
            is_free=True,
            is_acapulco=True,
            state_key='guerrero',
            state_label='Guerrero',
            zone_key='acapulco',
            label='Envío gratis a Acapulco',
        )

    state_key = resolve_state_key(state)
    if state_key:
        fee = STATE_SHIPPING_RATES[state_key]
        state_label = STATE_LABELS[state_key]
        return ShippingQuote(
            fee=fee,
            is_free=False,
            is_acapulco=False,
            state_key=state_key,
            state_label=state_label,
            zone_key=state_key,
            label=f'Envío a {state_label}: ${fee:,.2f}',
        )

    return ShippingQuote(
        fee=DEFAULT_NATIONAL_RATE,
        is_free=False,
        is_acapulco=False,
        state_key='',
        state_label='',
        zone_key='nacional',
        label=f'Envío nacional: ${DEFAULT_NATIONAL_RATE:,.2f}',
    )

File: MCD-Agencia-main/frontend/src/lib/shipping.ts
/**
 * Shipping rates from Acapulco, Guerrero to the rest of Mexico.
 * Keep in sync with backend/apps/orders/services/shipping.py
 */

export const MEXICO_STATE_OPTIONS = [
  { value: 'Aguascalientes', label: 'Aguascalientes' },
  { value: 'Baja California', label: 'Baja California' },
  { value: 'Baja California Sur', label: 'Baja California Sur' },
  { value: 'Campeche', label: 'Campeche' },
  { value: 'Chiapas', label: 'Chiapas' },
  { value: 'Chihuahua', label: 'Chihuahua' },
  { value: 'Ciudad de México', label: 'Ciudad de México' },
  { value: 'Coahuila', label: 'Coahuila' },
  { value: 'Colima', label: 'Colima' },
  { value: 'Durango', label: 'Durango' },
  { value: 'Estado de México', label: 'Estado de México' },
  { value: 'Guanajuato', label: 'Guanajuato' },
  { value: 'Guerrero', label: 'Guerrero' },
  { value: 'Hidalgo', label: 'Hidalgo' },
  { value: 'Jalisco', label: 'Jalisco' },
  { value: 'Michoacán', label: 'Michoacán' },
  { value: 'Morelos', label: 'Morelos' },
  { value: 'Nayarit', label: 'Nayarit' },
  { value: 'Nuevo León', label: 'Nuevo León' },
  { value: 'Oaxaca', label: 'Oaxaca' },
  { value: 'Puebla', label: 'Puebla' },
  { value: 'Querétaro', label: 'Querétaro' },
  { value: 'Quintana Roo', label: 'Quintana Roo' },
  { value: 'San Luis Potosí', label: 'San Luis Potosí' },
  { value: 'Sinaloa', label: 'Sinaloa' },
  { value: 'Sonora', label: 'Sonora' },
  { value: 'Tabasco', label: 'Tabasco' },
  { value: 'Tamaulipas', label: 'Tamaulipas' },
  { value: 'Tlaxcala', label: 'Tlaxcala' },
  { value: 'Veracruz', label: 'Veracruz' },
  { value: 'Yucatán', label: 'Yucatán' },
  { value: 'Zacatecas', label: 'Zacatecas' },
] as const;

const STATE_SHIPPING_RATES: Record<string, number> = {
  aguascalientes: 289,
  'baja california': 499,
  'baja california sur': 529,
  campeche: 369,
  chiapas: 329,
  chihuahua: 449,
  'ciudad de mexico': 199,
  coahuila: 389,
  colima: 229,
  durango: 349,
  'estado de mexico': 199,
  guanajuato: 269,
  guerrero: 99,
  hidalgo: 219,
  jalisco: 269,
  michoacan: 199,
  morelos: 169,
  nayarit: 289,
  'nuevo leon': 379,
  oaxaca: 149,
  puebla: 189,
  queretaro: 249,
  'quintana roo': 429,
  'san luis potosi': 299,
  sinaloa: 399,
  sonora: 459,
  tabasco: 319,
  tamaulipas: 359,
  tlaxcala: 209,
  veracruz: 219,
  yucatan: 419,
  zacatecas: 319,
};

const STATE_LABELS: Record<string, string> = {
  aguascalientes: 'Aguascalientes',
  'baja california': 'Baja California',
  'baja california sur': 'Baja California Sur',
  campeche: 'Campeche',
  chiapas: 'Chiapas',
  chihuahua: 'Chihuahua',
  'ciudad de mexico': 'Ciudad de México',
  coahuila: 'Coahuila',
  colima: 'Colima',
  durango: 'Durango',
  'estado de mexico': 'Estado de México',
  guanajuato: 'Guanajuato',
  guerrero: 'Guerrero',
  hidalgo: 'Hidalgo',
  jalisco: 'Jalisco',
  michoacan: 'Michoacán',
  morelos: 'Morelos',
  nayarit: 'Nayarit',
  'nuevo leon': 'Nuevo León',
  oaxaca: 'Oaxaca',
  puebla: 'Puebla',
  queretaro: 'Querétaro',
  'quintana roo': 'Quintana Roo',
  'san luis potosi': 'San Luis Potosí',
  sinaloa: 'Sinaloa',
  sonora: 'Sonora',
  tabasco: 'Tabasco',
  tamaulipas: 'Tamaulipas',
  tlaxcala: 'Tlaxcala',
  veracruz: 'Veracruz',
  yucatan: 'Yucatán',
  zacatecas: 'Zacatecas',
};

const STATE_ALIASES: Record<string, string> = {
  ags: 'aguascalientes',
  bc: 'baja california',
  bcs: 'baja california sur',
  camp: 'campeche',
  chis: 'chiapas',
  chih: 'chihuahua',
  cdmx: 'ciudad de mexico',
  df: 'ciudad de mexico',
  'd f': 'ciudad de mexico',
  'distrito federal': 'ciudad de mexico',
  'ciudad mexico': 'ciudad de mexico',
  'mexico city': 'ciudad de mexico',
  coah: 'coahuila',
  'coahuila de zaragoza': 'coahuila',
  col: 'colima',
  dgo: 'durango',
  edomex: 'estado de mexico',
  'edo de mexico': 'estado de mexico',
  'edo mexico': 'estado de mexico',
  'estado mexico': 'estado de mexico',
  mex: 'estado de mexico',
  mexico: 'estado de mexico',
  gto: 'guanajuato',
  gro: 'guerrero',
  hgo: 'hidalgo',
  jal: 'jalisco',
  mich: 'michoacan',
  mor: 'morelos',
  nay: 'nayarit',
  nl: 'nuevo leon',
  'n l': 'nuevo leon',
  oax: 'oaxaca',
  pue: 'puebla',
  qro: 'queretaro',
  'q roo': 'quintana roo',
  qroo: 'quintana roo',
  qr: 'quintana roo',
  slp: 'san luis potosi',
  sin: 'sinaloa',
  son: 'sonora',
  tab: 'tabasco',
  tamps: 'tamaulipas',
  tam: 'tamaulipas',
  tlax: 'tlaxcala',
  ver: 'veracruz',
  'veracruz de ignacio de la llave': 'veracruz',
  yuc: 'yucatan',
  zac: 'zacatecas',
};

const ACAPULCO_POSTAL_PREFIXES = new Set(['393', '394', '395', '396', '397', '398', '399']);
const DEFAULT_NATIONAL_RATE = 299;

export type ShippingQuote = {
  fee: number;
  isFree: boolean;
  isAcapulco: boolean;
  stateKey: string;
  stateLabel: string;
  zoneKey: string;
  label: string;
};

function normalize(value: string): string {
  return (value || '')
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .replace(/[^a-z0-9\s]/g, ' ')
    .replace(/\s+/g, ' ')
    .trim();
}

function resolveStateKey(state: string): string {
  const normalized = normalize(state);
  if (!normalized) return '';
  if (normalized in STATE_SHIPPING_RATES) return normalized;
  if (normalized in STATE_ALIASES) return STATE_ALIASES[normalized];
  for (const [key, label] of Object.entries(STATE_LABELS)) {
    if (normalize(label) === normalized) return key;
  }
  return '';
}

function isAcapulcoDestination(city = '', state = '', postalCode = ''): boolean {
  const cityName = normalize(city);
  if (cityName) {
    return cityName.includes('acapulco');
  }
  const digits = (postalCode || '').replace(/\D/g, '');
  if (digits.length >= 5 && ACAPULCO_POSTAL_PREFIXES.has(digits.slice(0, 3))) {
    const stateKey = resolveStateKey(state);
    return !stateKey || stateKey === 'guerrero';
  }
  return false;
}

export function calculateShippingQuote(input: {
  city?: string;
  state?: string;
  postal_code?: string;
}): ShippingQuote {
  const city = input.city || '';
  const state = input.state || '';
  const postalCode = input.postal_code || '';

  if (isAcapulcoDestination(city, state, postalCode)) {
    return {
      fee: 0,
      isFree: true,
      isAcapulco: true,
      stateKey: 'guerrero',
      stateLabel: 'Guerrero',
      zoneKey: 'acapulco',
      label: 'Envío gratis a Acapulco',
    };
  }

  const stateKey = resolveStateKey(state);
  if (stateKey) {
    const fee = STATE_SHIPPING_RATES[stateKey];
    const stateLabel = STATE_LABELS[stateKey];
    return {
      fee,
      isFree: false,
      isAcapulco: false,
      stateKey,
      stateLabel,
      zoneKey: stateKey,
      label: `Envío a ${stateLabel}`,
    };
  }

  return {
    fee: DEFAULT_NATIONAL_RATE,
    isFree: false,
    isAcapulco: false,
    stateKey: '',
    stateLabel: '',
    zoneKey: 'nacional',
    label: 'Envío nacional',
  };
}
